core/src/test/scala/org/apache/spark/storage/BlockManagerInfoSuite.scala

```diff
@@ -0,0 +1,157 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.storage
+
+import java.util.{HashMap => JHashMap}
+
+import scala.collection.JavaConverters._
+
+import org.apache.spark.SparkFunSuite
+
+class BlockManagerInfoSuite extends SparkFunSuite {
+
+  def testWithShuffleServiceOnOff(testName: String)
+      (f: (Boolean, BlockManagerInfo) => Unit): Unit = {
+    Seq(true, false).foreach { svcEnabled =>
+      val bmInfo = new BlockManagerInfo(
+        BlockManagerId("executor0", "host", 1234, None),
+        timeMs = 300,
+        maxOnHeapMem = 10000,
+        maxOffHeapMem = 20000,
+        slaveEndpoint = null,
+        if (svcEnabled) Some(new JHashMap[BlockId, BlockStatus]) else None)
+      test(s"$testName externalShuffleServiceEnabled=$svcEnabled") {
+        f(svcEnabled, bmInfo)
+      }
+    }
+  }
+
+  testWithShuffleServiceOnOff("broadcast block") { (_, bmInfo) =>
+    val broadcastId: BlockId = BroadcastBlockId(0, "field1")
+    bmInfo.updateBlockInfo(
+      broadcastId, StorageLevel.MEMORY_AND_DISK, memSize = 200, diskSize = 100)
+    assert(bmInfo.blocks.asScala ===
+      Map(broadcastId -> BlockStatus(StorageLevel.MEMORY_AND_DISK, 0, 100)))
+    assert(bmInfo.exclusiveCachedBlocks.isEmpty)
+    assert(bmInfo.remainingMem === 29800)
+  }
+
+  testWithShuffleServiceOnOff("RDD block with MEMORY_ONLY") { (svcEnabled, bmInfo) =>
+    val rddId: BlockId = RDDBlockId(0, 0)
+    bmInfo.updateBlockInfo(rddId, StorageLevel.MEMORY_ONLY, memSize = 200, diskSize = 0)
+    assert(bmInfo.blocks.asScala ===
+      Map(rddId -> BlockStatus(StorageLevel.MEMORY_ONLY, 200, 0)))
+    assert(bmInfo.exclusiveCachedBlocks === Set(rddId))
+    assert(bmInfo.remainingMem === 29800)
+    if (svcEnabled) {
+      assert(bmInfo.externalShuffleServiceBlockStatus.get.isEmpty)
+    }
+  }
+
+  testWithShuffleServiceOnOff("RDD block with MEMORY_AND_DISK") { (svcEnabled, bmInfo) =>
+    val rddId: BlockId = RDDBlockId(0, 0)
+    bmInfo.updateBlockInfo(
+      rddId, StorageLevel.MEMORY_AND_DISK, memSize = 200, diskSize = 400)
```

Comment:
this is the effective storage level, not the requested level, right?  So this is testing that a block actually exists in both memory and disk -- not that the user called `rdd.persist(StorageLevel.MEMORY_AND_DISK)`, which usually only stores the data in one of memory or disk, but not both.

As both marcelo and I asked about it -- are there any tests for when the requested storage level is memory and disk, but you might get either one?  Your walk through of how it works in the code is good, just wondering if there are tests.

Reply:
Yes this is the effective storage level.

Actually as I see an effective storage level can never ever be `StorageLevel.MEMORY_AND_DISK`. 
In my walk through we reached the conclusion that always the `BlockManager#getCurrentBlockStatus` is used for computing the effective storage level (or it could also come from `BlockStatus#empty` which means `StorageLevel$#NONE` so that case is not relevant). So in `getCurrentBlockStatus` the level is calculated from the requested level which there referenced as `level`, and the related code:

https://github.com/apache/spark/blob/3e7797aa02a48e653b3c167a62cba6512a1d3bcb/core/src/main/scala/org/apache/spark/storage/BlockManager.scala#L691-L701

You are right a block either can be found in the memory store or in the disk store (I have double checked it by focusing to  how `BlockManager#doPutIterator` uses the result of  `MemoryStore#putIteratorAsValues`). 

So in the `getCurrentBlockStatus` either  `inMem` or `onDisk` is true:

```scala
   val inMem = level.useMemory && memoryStore.contains(blockId) 
   val onDisk = level.useDisk && diskStore.contains(blockId) 
```

Based on the analysis above I would delete this test with with `MEMORY_AND_DISK`.

Reply:
yes I was wondering the same thing -- I agree the effective storage level can't be both memory and disk on initial storage.  I hadn't confirmed whether it could never become both memory and disk eventually, though -- eg. there isn't space initially, so it gets written to disk, but then later on other RDDs are unpersisted, but the RDD is accessed again, so it gets promoted to being in memory, and the files on disk are still there.  See the `maybeCacheDiskBytesInMemory` / `maybeCacheDiskValuesInMemory`: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/storage/BlockManager.scala#L774-L776

so the test case may still be useful, but its not actually the more important part of a _requested_  MEMORY_AND_DISK storage level that marcelo and I were concerned about

Reply:
I see. And for example a reregister later on will lead to `BlockManager#reportAllBlocks` where `getCurrentBlockStatus` will see this block both in the memory store and in the disk store. Then I add this test back.